### PR TITLE
add no-useless-intersection rule

### LINKED ISSUE #314: Rule: Intersection with type without members
Some intersection types plainly don't make sense as there's nothing that can satisfy them :
```
function foo(p: number & null) {
  ...
}
```

### PR DESCRIPTION
**Pull request by stas-vilchik**
fixed #314

### CODE REVIEW

File: sonarts-core/src/utils/navigation.ts [hunk 1]
```diff
@@ -136,6 +136,14 @@ function isToken(node: ts.Node): boolean {
   return node.kind <= ts.SyntaxKind.OfKeyword;
 }
 
+export function isIntersectionTypeNode(node: ts.Node): node is ts.IntersectionTypeNode {
```
Comment (vilchik-elena) [inline sonarts-core/src/utils/navigation.ts]: @stas-vilchik will we create a function for every node? IMO there is no need for that

File: sonarts-core/src/rules/noUselessIntersectionRule.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+function isTypeWithoutMembers(type: ts.Type) {
+  const isNullLike = isNullType(type) || isUndefinedType(type) || isVoidType(type);
+  const isEmptyInterface = Boolean(
+    type.symbol && type.symbol.members && type.symbol.members.size === 0 && isStandaloneInterface(type.symbol),
+  );
+  return isNullLike || isEmptyInterface;
+}
+
+function isStandaloneInterface({ declarations }: ts.Symbol) {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noUselessIntersectionRule.ts]: @stas-vilchik what does mean standalone interface?

File: sonarts-core/src/rules/noUselessIntersectionRule.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  // there is no declarations for `{}`
+  // otherwise check that none of declarations has a heritage clause (`extends X` or `implments X`)
+  return (
+    !declarations ||
+    declarations.every(declaration => {
```
Comment (vilchik-elena) [inline sonarts-core/src/rules/noUselessIntersectionRule.ts]: @stas-vilchik how one symbol could have several declarations?

Comment (stas-vilchik) [reply sonarts-core/src/rules/noUselessIntersectionRule.ts]: I can think only about function overloads

File: sonarts-core/src/utils/navigation.ts (hunk 1 above)
Comment (stas-vilchik) [reply sonarts-core/src/utils/navigation.ts]: ???

File: sonarts-core/src/rules/noUselessIntersectionRule.ts (hunk 2 above)
Comment (stas-vilchik) [reply sonarts-core/src/rules/noUselessIntersectionRule.ts]: Interface which doesn't extend anything. I couldn't find a better name, do you have any ideas?

Comment (vilchik-elena) [reply sonarts-core/src/rules/noUselessIntersectionRule.ts]: @stas-vilchik `isMemberless`?:)

**Pull request merged**